subiquity/server/controllers/install.py

```diff
@@ -222,6 +222,51 @@ async def run_curtin_step(
                 device_map = json.load(fp)
             self.app.controllers.Filesystem.update_devices(device_map)
 
+    @with_context(description="configuring OEM packages")
+    async def pre_curthooks_oem_configuration(self, context):
+        # For OEM, we basically mimic what ubuntu-drivers does:
+        # 1. Install each package with apt-get install
+        # 2. For each package, run apt-get update using only the source
+        # installed by said package.
+        # 3. Run apt-get install again for each package. This will upgrade
+        # them to the version found in the OEM archive.
+
+        # NOTE In ubuntu-drivers, this is done in a single call to apt-get
+        # install.
+        for pkg in self.model.oem.metapkgs:
+            await self.install_package(package=pkg.name)
+
+        if self.model.network.has_network:
+            for pkg in self.model.oem.metapkgs:
+                source_list = f"/etc/apt/sources.list.d/{pkg.name}.list"
+                await run_curtin_command(
+                    self.app,
+                    context,
+                    "in-target",
+                    "-t",
+                    self.tpath(),
+                    "--",
+                    "apt-get",
+                    "update",
+                    "-o",
+                    f"Dir::Etc::SourceList={source_list}",
+                    "-o",
+                    "Dir::Etc::SourceParts=/dev/null",
+                    "--no-list-cleanup",
+                    private_mounts=False,
+                )
+
+            # NOTE In ubuntu-drivers, this is done in a single call to
+            # apt-get install.
+            for pkg in self.model.oem.metapkgs:
+                await self.install_package(package=pkg.name)
+
+        # If we already have a kernel installed, don't bother requesting
+        # curthooks to install it again or we might end up with two
+        # kernels.
+        if await list_installed_kernels(Path(self.tpath())):
+            self.model.kernel.curthooks_no_install = True
```

Comment:
I think these lines should go into a separate method too.

Reply:
Because it can also have an effect outside the OEM scenario - when the source squashfs contains a kernel package (it was the case until recently on desktop). See https://bugs.launchpad.net/subiquity/+bug/2026225

Whether it's a behavior we want or not is a different question. But I'd rather keep the current behavior (i.e., run it even if no OEM metapackages are applicable) at least for this PR - if that makes sense?

Reply:
Oh I see. I pushed something like that.